python/tests/catalog/test_hive.py

```diff
@@ -0,0 +1,339 @@
+        id=None,
+        fileMetadata=None,
+        dictionary=None,
+        txnId=None,
+    )
+
+    catalog.load_table(("default", "table"))
+
+
+def test_rename_table_from_does_not_exists():
+    catalog = HiveCatalog(HIVE_CATALOG_NAME, {}, url=HIVE_METASTORE_FAKE_URL)
+
+    catalog._client = MagicMock()
+    catalog._client.__enter__().alter_table.side_effect = NoSuchObjectException(
+        message="hive.default.does_not_exists table not found"
+    )
+
+    with pytest.raises(NoSuchTableError) as exc_info:
+        catalog.rename_table(("default", "does_not_exists"), ("default", "new_table"))
+
+    assert "Table default.does_not_exists not found" in str(exc_info.value)
+
+
+def test_rename_table_to_namespace_does_not_exists():
+    catalog = HiveCatalog(HIVE_CATALOG_NAME, {}, url=HIVE_METASTORE_FAKE_URL)
+
+    catalog._client = MagicMock()
+    catalog._client.__enter__().alter_table.side_effect = InvalidOperationException(
+        message="Unable to change partition or table. Database default does not exist Check metastore logs for detailed stack.does_not_exists"
+    )
+
+    with pytest.raises(NoSuchNamespaceError) as exc_info:
+        catalog.rename_table(("default", "does_exists"), ("default_does_not_exists", "new_table"))
+
+    assert "Destination database default_does_not_exists does not exists" in str(exc_info.value)
+
+
+def test_drop_database_does_not_empty():
+    catalog = HiveCatalog(HIVE_CATALOG_NAME, {}, url=HIVE_METASTORE_FAKE_URL)
+
+    catalog._client = MagicMock()
+    catalog._client.__enter__().drop_database.side_effect = InvalidOperationException(
+        message="Database not_empty is not empty. One or more tables exist."
+    )
+
+    with pytest.raises(NamespaceNotEmptyError) as exc_info:
+        catalog.drop_namespace(("not_empty",))
+
+    assert "Database not_empty is not empty" in str(exc_info.value)
+
+
+def test_drop_database_does_not_exists():
+    catalog = HiveCatalog(HIVE_CATALOG_NAME, {}, url=HIVE_METASTORE_FAKE_URL)
+
+    catalog._client = MagicMock()
+    catalog._client.__enter__().drop_database.side_effect = MetaException(message="java.lang.NullPointerException")
+
+    with pytest.raises(NoSuchNamespaceError) as exc_info:
+        catalog.drop_namespace(("does_not_exists",))
+
+    assert "Database does_not_exists does not exists" in str(exc_info.value)
+
+
+def test_list_tables():
+    catalog = HiveCatalog(HIVE_CATALOG_NAME, {}, url=HIVE_METASTORE_FAKE_URL)
+
+    catalog._client = MagicMock()
+    catalog._client.__enter__().get_all_tables.return_value = ["table1", "table2"]
+
+    assert catalog.list_tables("database") == ["table1", "table2"]
+
+
+def test_list_namespaces():
+    catalog = HiveCatalog(HIVE_CATALOG_NAME, {}, url=HIVE_METASTORE_FAKE_URL)
+
+    catalog._client = MagicMock()
+    catalog._client.__enter__().get_all_databases.return_value = ["namespace1", "namespace2"]
+
+    assert catalog.list_namespaces() == ["namespace1", "namespace2"]
```

Comment:
This should be `[("namespace1",), ("namespace2",)]` because it is a list of identifiers.

Reply:
I'm curious: why didn't mypy catch this?

Reply:
Great catch, thanks. The generated thrift code doesn't have any type annotations at all, so it can't make any assumptions about what it returns. Next to that, mypy isn't smart enough to check that the assertion doesn't make any sense (since `catalog.list_namespaces()` does have an annotation).